File: iosApp/iosApp/Modules/Dashboard/DashboardWireframeFactory.swift
// Created by web3d3v on 11/02/2022.
// Copyright (c) 2022 Sons Of Crypto.
// SPDX-License-Identifier: MIT

import UIKit
import web3walletcore

// MARK: - DashboardWireframeFactory

protocol DashboardWireframeFactory: AnyObject {
    func make(_ parent: UIViewController?) -> DashboardWireframe
}

// MARK: - DefaultDashboardWireframeFactory

final class DefaultDashboardWireframeFactory {
    private let accountWireframeFactory: AccountWireframeFactory
    private let alertWireframeFactory: AlertWireframeFactory
    private let mnemonicConfirmationWireframeFactory: MnemonicConfirmationWireframeFactory
    private let currencyPickerWireframeFactory: CurrencyPickerWireframeFactory
    private let currencyReceiveWireframeFactory: CurrencyReceiveWireframeFactory
    private let currencySendWireframeFactory: CurrencyCurrencyWireframeFactory
    private let currencySwapWireframeFactory: CurrencySwapWireframeFactory
    private let nftDetailWireframeFactory: NFTDetailWireframeFactory
    private let qrCodeScanWireframeFactory: QRCodeScanWireframeFactory
    private let themePickerWireframeFactory: ThemePickerWireframeFactory
    private let improvementProposalsWireframeFactory: ImprovementProposalsWireframeFactory
    private let networksService: NetworksService
    private let currencyStoreService: CurrencyStoreService
    private let walletService: WalletService
    private let nftsService: NFTsService
    private let actionsService: ActionsService

    init(
        accountWireframeFactory: AccountWireframeFactory,
        alertWireframeFactory: AlertWireframeFactory,
        mnemonicConfirmationWireframeFactory: MnemonicConfirmationWireframeFactory,
        currencyPickerWireframeFactory: CurrencyPickerWireframeFactory,
        currencyReceiveWireframeFactory: CurrencyReceiveWireframeFactory,
        currencySendWireframeFactory: CurrencyCurrencyWireframeFactory,
        currencySwapWireframeFactory: CurrencySwapWireframeFactory,
        nftDetailWireframeFactory: NFTDetailWireframeFactory,
        qrCodeScanWireframeFactory: QRCodeScanWireframeFactory,
        themePickerWireframeFactory: ThemePickerWireframeFactory,
        improvementProposalsWireframeFactory: ImprovementProposalsWireframeFactory,
        networksService: NetworksService,
        currencyStoreService: CurrencyStoreService,
        walletService: WalletService,
        nftsService: NFTsService,
        actionsService: ActionsService
    ) {
        self.accountWireframeFactory = accountWireframeFactory
        self.alertWireframeFactory = alertWireframeFactory
        self.mnemonicConfirmationWireframeFactory = mnemonicConfirmationWireframeFactory
        self.currencyPickerWireframeFactory = currencyPickerWireframeFactory
        self.currencyReceiveWireframeFactory = currencyReceiveWireframeFactory
        self.currencySendWireframeFactory = currencySendWireframeFactory
        self.currencySwapWireframeFactory = currencySwapWireframeFactory
        self.nftDetailWireframeFactory = nftDetailWireframeFactory
        self.qrCodeScanWireframeFactory = qrCodeScanWireframeFactory
        self.themePickerWireframeFactory = themePickerWireframeFactory
        self.improvementProposalsWireframeFactory = improvementProposalsWireframeFactory
        self.networksService = networksService
        self.currencyStoreService = currencyStoreService
        self.walletService = walletService
        self.nftsService = nftsService
        self.actionsService = actionsService
    }
}

extension DefaultDashboardWireframeFactory: DashboardWireframeFactory {

    func make(_ parent: UIViewController?) -> DashboardWireframe {
        DefaultDashboardWireframe(
            parent,
            accountWireframeFactory: accountWireframeFactory,
            alertWireframeFactory: alertWireframeFactory,
            mnemonicConfirmationWireframeFactory: mnemonicConfirmationWireframeFactory,
            currencyPickerWireframeFactory: currencyPickerWireframeFactory,
            currencyReceiveWireframeFactory: currencyReceiveWireframeFactory,
            currencySendWireframeFactory: currencySendWireframeFactory,
            currencySwapWireframeFactory: currencySwapWireframeFactory,
            nftDetailWireframeFactory: nftDetailWireframeFactory,
            qrCodeScanWireframeFactory: qrCodeScanWireframeFactory,
            themePickerWireframeFactory: themePickerWireframeFactory,
            improvementProposalsWireframeFactory: improvementProposalsWireframeFactory,
            networksService: networksService,
            currencyStoreService: currencyStoreService,
            walletService: walletService,
            nftsService: nftsService,
            actionsService: actionsService
        )
    }
}

// MARK: - Assembler

final class DashboardWireframeFactoryAssembler: AssemblerComponent {

    func register(to registry: AssemblerRegistry) {
        registry.register(scope: .instance) { resolver -> DashboardWireframeFactory in
            DefaultDashboardWireframeFactory(
                accountWireframeFactory: resolver.resolve(),
                alertWireframeFactory: resolver.resolve(),
                mnemonicConfirmationWireframeFactory: resolver.resolve(),
                currencyPickerWireframeFactory: resolver.resolve(),
                currencyReceiveWireframeFactory: resolver.resolve(),
                currencySendWireframeFactory: resolver.resolve(),
                currencySwapWireframeFactory: resolver.resolve(),
                nftDetailWireframeFactory: resolver.resolve(),
                qrCodeScanWireframeFactory: resolver.resolve(),
                themePickerWireframeFactory: resolver.resolve(),
                improvementProposalsWireframeFactory: resolver.resolve(),
                networksService: resolver.resolve(),
                currencyStoreService: resolver.resolve(),
                walletService: resolver.resolve(),
                nftsService: resolver.resolve(),
                actionsService: resolver.resolve()
            )
        }
    }
}

File: androidApp/src/main/java/com/sonsofcrypto/web3wallet/android/modules/compose/dashboard/DashboardWireframeFactory.kt
package com.sonsofcrypto.web3wallet.android.modules.compose.dashboard

import androidx.fragment.app.Fragment
import com.sonsofcrypto.web3lib.services.currencyStore.CurrencyStoreService
import com.sonsofcrypto.web3lib.services.networks.NetworksService
import com.sonsofcrypto.web3lib.services.wallet.WalletService
import com.sonsofcrypto.web3lib.utils.WeakRef
import com.sonsofcrypto.web3wallet.android.common.AssemblerComponent
import com.sonsofcrypto.web3wallet.android.common.AssemblerRegistry
import com.sonsofcrypto.web3wallet.android.common.AssemblerRegistryScope
import com.sonsofcrypto.web3wallet.android.modules.compose.account.AccountWireframeFactory
import com.sonsofcrypto.web3wallet.android.modules.compose.currencypicker.CurrencyPickerWireframeFactory
import com.sonsofcrypto.web3wallet.android.modules.compose.currencyreceive.CurrencyReceiveWireframeFactory
import com.sonsofcrypto.web3wallet.android.modules.compose.currencysend.CurrencySendWireframeFactory
import com.sonsofcrypto.web3wallet.android.modules.compose.currencyswap.CurrencySwapWireframeFactory
import com.sonsofcrypto.web3wallet.android.modules.compose.improvementproposals.ImprovementProposalsWireframeFactory
import com.sonsofcrypto.web3wallet.android.modules.compose.mnemonicconfirmation.MnemonicConfirmationWireframeFactory
import com.sonsofcrypto.web3wallet.android.modules.compose.nftdetail.NFTDetailWireframeFactory
import com.sonsofcrypto.web3wallet.android.modules.compose.qrcodescan.QRCodeScanWireframeFactory
import com.sonsofcrypto.web3walletcore.modules.dashboard.DashboardWireframe
import com.sonsofcrypto.web3walletcore.services.actions.ActionsService
import com.sonsofcrypto.web3walletcore.services.nfts.NFTsService
import smartadapter.internal.extension.name

interface DashboardWireframeFactory {
    fun make(parent: Fragment?): DashboardWireframe
}

class DefaultDashboardWireframeFactory(
    private val networksService: NetworksService,
    private val currencyStoreService: CurrencyStoreService,
    private val walletService: WalletService,
    private val nftsService: NFTsService,
    private val actionsService: ActionsService,
    private val currencyReceiveWireframeFactory: CurrencyReceiveWireframeFactory,
    private val currencySendWireframeFactory: CurrencySendWireframeFactory,
    private val currencySwapWireframeFactory: CurrencySwapWireframeFactory,
    private val currencyPickerWireframeFactory: CurrencyPickerWireframeFactory,
    private val accountWireframeFactory: AccountWireframeFactory,
    private val nftDetailWireframeFactory: NFTDetailWireframeFactory,
    private val mnemonicConfirmationWireframeFactory: MnemonicConfirmationWireframeFactory,
    private val improvementProposalsWireframeFactory: ImprovementProposalsWireframeFactory,
    private val qrCodeScanWireframeFactory: QRCodeScanWireframeFactory,
): DashboardWireframeFactory {

    override fun make(parent: Fragment?): DashboardWireframe = DefaultDashboardWireframe(
        parent?.let { WeakRef(it) },
        networksService,
        currencyStoreService,
        walletService,
        nftsService,
        actionsService,
        currencyReceiveWireframeFactory,
        currencySendWireframeFactory,
        currencySwapWireframeFactory,
        currencyPickerWireframeFactory,
        accountWireframeFactory,
        nftDetailWireframeFactory,
        mnemonicConfirmationWireframeFactory,
        improvementProposalsWireframeFactory,
        qrCodeScanWireframeFactory,
    )
}

class DashboardWireframeFactoryAssembler: AssemblerComponent {

    override fun register(to: AssemblerRegistry) {

        to.register(DashboardWireframeFactory::class.name, AssemblerRegistryScope.INSTANCE) {
            DefaultDashboardWireframeFactory(
                it.resolve(NetworksService::class.name),
                it.resolve(CurrencyStoreService::class.name),
                it.resolve(WalletService::class.name),
                it.resolve(NFTsService::class.name),
                it.resolve(ActionsService::class.name),
                it.resolve(CurrencyReceiveWireframeFactory::class.name),
                it.resolve(CurrencySendWireframeFactory::class.name),
                it.resolve(CurrencySwapWireframeFactory::class.name),
                it.resolve(CurrencyPickerWireframeFactory::class.name),
                it.resolve(AccountWireframeFactory::class.name),
                it.resolve(NFTDetailWireframeFactory::class.name),
                it.resolve(MnemonicConfirmationWireframeFactory::class.name),
                it.resolve(ImprovementProposalsWireframeFactory::class.name),
                it.resolve(QRCodeScanWireframeFactory::class.name),
            )
        }
    }
}